sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpResponseMessage.cs

```diff
@@ -69,8 +69,13 @@ public IEnumerable<TValue> GetListValue<TValue>(MapKey key)
             }
 
             var list = (List<object>)Map[key];
-
-            return list.Cast<TValue>();
+            List<TValue> values = null;
+            foreach (var item in list)
```

Comment:
I wonder, do we save enough with this change that the intermediary list instantiation is worth it to get the count?   We could do a `yield return` here and use an `out` param for the length, I think....

Reply:
We actually we would save something by changing the return type to the list and then there would be no enumerator allocation anymore. I initially wanted to go with that but then I remembered your enumerable all the things rule so I dismissed it.

But yeah yield return and counter would work too but still allocate the enumerator right?

Reply:
I think it's safe to return this as a list-type here, as it's an internal class.  We try to keep the public API as generic as we can to avoid boxing ourselves in with too-specific a contract.   Nothing breaks if we try to change this later.

@JoshLove-msft:  Any objections to potentially changing the return type here?

Reply:
> We could do a yield return here and use an out param for the length, I think....

iterators cannot have out parameters unfortunately

Reply:
Did a quick benchmark

```
   [Config(typeof(Config))]
    public class EnumerableCount
    {
        private List<object> data;
        private Consumer consumer;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(1000000));
            }
        }
        [IterationSetup]
        public void Setup()
        {
            data = Enumerable.Repeat(new SomeClass(), NumberOfItems).Cast<object>().ToList();
            consumer = new Consumer();
        }

        [Params(0, 1, 5)]
        public int NumberOfItems { get; set; }

        class SomeClass {}

        [Benchmark(Baseline = true)]
        public void Original()
        {
            GetValue<SomeClass>().Consume(consumer);
        }

        IEnumerable<TValue> GetValue<TValue>()
        {
            return data.Cast<TValue>();
        }

        [Benchmark]
        public void List()
        {
            GetValueList<SomeClass>().Consume(consumer);
        }

        IEnumerable<TValue> GetValueList<TValue>()
        {
            List<TValue> values = null;
            foreach (var item in data)
            {
                values ??= new List<TValue>(data.Count);
                values.Add((TValue)item);
            }
            return values ?? Enumerable.Empty<TValue>();
        }

        [Benchmark]
        public void ListReturnList()
        {
            GetValueListReturnList<SomeClass>().Consume(consumer);
        }

        List<TValue> GetValueListReturnList<TValue>()
        {
            var values = new List<TValue>(data.Count);
            foreach (var item in data)
            {
                values.Add((TValue)item);
            }
            return values;
        }
    }
```


``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.202
  [Host]     : .NET Core 3.1.14 (CoreCLR 4.700.21.16201, CoreFX 4.700.21.16208), X64 RyuJIT
  Job-UZLRLE : .NET Core 3.1.14 (CoreCLR 4.700.21.16201, CoreFX 4.700.21.16208), X64 RyuJIT

InvocationCount=1000000  

```
|         Method | NumberOfItems |      Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |----------:|---------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|       **Original** |             **0** | **117.15 ns** | **0.907 ns** | **0.757 ns** |  **1.00** |    **0.00** | **0.0110** |     **-** |     **-** |      **96 B** |
|           List |             0 |  17.74 ns | 0.226 ns | 0.278 ns |  0.15 |    0.00 |      - |     - |     - |         - |
| ListReturnList |             0 |  32.90 ns | 0.561 ns | 0.623 ns |  0.28 |    0.01 | 0.0080 |     - |     - |      72 B |
|                |               |           |          |          |       |         |        |       |       |           |
|       **Original** |             **1** | **150.39 ns** | **1.639 ns** | **1.533 ns** |  **1.00** |    **0.00** | **0.0110** |     **-** |     **-** |      **96 B** |
|           List |             1 |  56.27 ns | 1.150 ns | 1.412 ns |  0.38 |    0.01 | 0.0120 |     - |     - |     104 B |
| ListReturnList |             1 |  51.97 ns | 0.755 ns | 0.706 ns |  0.35 |    0.00 | 0.0120 |     - |     - |     104 B |
|                |               |           |          |          |       |         |        |       |       |           |
|       **Original** |             **5** | **241.15 ns** | **1.653 ns** | **1.466 ns** |  **1.00** |    **0.00** | **0.0110** |     **-** |     **-** |      **96 B** |
|           List |             5 | 140.63 ns | 1.980 ns | 1.852 ns |  0.58 |    0.01 | 0.0160 |     - |     - |     136 B |
| ListReturnList |             5 | 141.84 ns | 2.861 ns | 5.443 ns |  0.56 |    0.02 | 0.0160 |     - |     - |     136 B |

Reply:
I'll let you decide if the changes are worth it or not. Also happy to close

Reply:
I can also try to compare the approaches in a benchmark that takes the growing into account. Let me know if you think it is worth the time

Reply:
Do we know why the List/ListReturnList allocate more than the Original (when there is at least 1 element)?

Reply:
Here is a much more realistic benachmark

```
    [Config(typeof(Config))]
    public class EnumerableCount
    {
        private List<object> data;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(160000));
            }
        }
        [IterationSetup]
        public void Setup()
        {
            data = Enumerable.Repeat(new SomeClass(), NumberOfItems).Cast<object>().ToList();
        }

        [Params(0, 1, 10, 32, 50, 100, 1000)]
        public int NumberOfItems { get; set; }

        public class SomeClass {}

        [Benchmark(Baseline = true)]
        public List<SomeClass> Original()
        {
            var copyList = new List<SomeClass>();
            foreach (var value in GetValue<SomeClass>())
            {
                copyList.Add(value);
            }

            return copyList;
        }

        IEnumerable<TValue> GetValue<TValue>()
        {
            return data.Cast<TValue>();
        }

        [Benchmark]
        public List<SomeClass> List()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueList<SomeClass>();
            foreach (var value in enumerable)
            {
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        IEnumerable<TValue> GetValueList<TValue>()
        {
            List<TValue> values = null;
            foreach (var item in data)
            {
                values ??= new List<TValue>(data.Count);
                values.Add((TValue)item);
            }
            return values ?? Enumerable.Empty<TValue>();
        }

        [Benchmark]
        public List<SomeClass> ListReturnList()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueListReturnList<SomeClass>();
            foreach (var value in enumerable)
            {
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        List<TValue> GetValueListReturnList<TValue>()
        {
            var values = new List<TValue>(data.Count);
            foreach (var item in data)
            {
                values.Add((TValue)item);
            }
            return values;
        }
    }
```

``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 3950X, 1 CPU, 32 logical and 16 physical cores
.NET Core SDK=5.0.202
  [Host]     : .NET Core 3.1.14 (CoreCLR 4.700.21.16201, CoreFX 4.700.21.16208), X64 RyuJIT
  Job-DZEQOL : .NET Core 3.1.14 (CoreCLR 4.700.21.16201, CoreFX 4.700.21.16208), X64 RyuJIT

InvocationCount=160000  

```
|         Method | NumberOfItems |         Mean |      Error |     StdDev | Ratio | RatioSD |  Gen 0 |  Gen 1 | Gen 2 | Allocated |
|--------------- |-------------- |-------------:|-----------:|-----------:|------:|--------:|-------:|-------:|------:|----------:|
|       **Original** |             **0** |    **109.79 ns** |   **1.966 ns** |   **1.642 ns** |  **1.00** |    **0.00** | **0.0125** |      **-** |     **-** |     **128 B** |
|           List |             0 |     20.56 ns |   0.257 ns |   0.228 ns |  0.19 |    0.00 |      - |      - |     - |         - |
| ListReturnList |             0 |     21.37 ns |   0.485 ns |   0.577 ns |  0.20 |    0.01 |      - |      - |     - |      32 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |             **1** |    **151.26 ns** |   **1.166 ns** |   **0.974 ns** |  **1.00** |    **0.00** | **0.0188** |      **-** |     **-** |     **184 B** |
|           List |             1 |    111.04 ns |   0.662 ns |   1.069 ns |  0.74 |    0.01 | 0.0188 |      - |     - |     168 B |
| ListReturnList |             1 |     56.15 ns |   1.182 ns |   1.161 ns |  0.37 |    0.01 | 0.0125 |      - |     - |     128 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |            **10** |    **408.22 ns** |   **5.716 ns** |   **5.347 ns** |  **1.00** |    **0.00** | **0.0500** |      **-** |     **-** |     **424 B** |
|           List |            10 |    286.76 ns |   5.144 ns |   4.811 ns |  0.70 |    0.01 | 0.0313 |      - |     - |     312 B |
| ListReturnList |            10 |    180.94 ns |   3.620 ns |   6.800 ns |  0.43 |    0.02 | 0.0313 |      - |     - |     272 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |            **32** |    **805.10 ns** |   **8.041 ns** |   **7.522 ns** |  **1.00** |    **0.00** | **0.0813** |      **-** |     **-** |     **704 B** |
|           List |            32 |    682.73 ns |  13.442 ns |  22.826 ns |  0.85 |    0.03 | 0.0750 |      - |     - |     664 B |
| ListReturnList |            32 |    480.99 ns |   8.732 ns |  11.354 ns |  0.60 |    0.02 | 0.0688 |      - |     - |     624 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |            **50** |  **1,215.02 ns** |  **15.337 ns** |  **12.807 ns** |  **1.00** |    **0.00** | **0.1438** |      **-** |     **-** |    **1240 B** |
|           List |            50 |    989.94 ns |  19.574 ns |  26.792 ns |  0.80 |    0.03 | 0.1125 |      - |     - |     952 B |
| ListReturnList |            50 |    767.92 ns |  15.339 ns |  29.554 ns |  0.64 |    0.02 | 0.1063 |      - |     - |     912 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |           **100** |  **2,324.06 ns** |  **46.462 ns** |  **65.133 ns** |  **1.00** |    **0.00** | **0.2688** |      **-** |     **-** |    **2288 B** |
|           List |           100 |  1,984.25 ns |  38.680 ns |  52.945 ns |  0.85 |    0.02 | 0.2063 |      - |     - |    1752 B |
| ListReturnList |           100 |  1,495.90 ns |  29.793 ns |  54.478 ns |  0.64 |    0.02 | 0.2000 |      - |     - |    1712 B |
|                |               |              |            |            |       |         |        |        |       |           |
|       **Original** |          **1000** | **19,816.56 ns** | **193.071 ns** | **180.598 ns** |  **1.00** |    **0.00** | **1.9938** | **0.1188** |     **-** |   **16696 B** |
|           List |          1000 | 18,498.34 ns | 214.753 ns | 179.328 ns |  0.93 |    0.01 | 1.9250 | 0.1063 |     - |   16152 B |
| ListReturnList |          1000 | 14,310.49 ns | 279.058 ns | 247.377 ns |  0.72 |    0.01 | 1.9188 | 0.1313 |     - |   16112 B |


Then we don't have to guess anymore and I think the results are pretty obvious

Reply:
For better display

![image](https://user-images.githubusercontent.com/174258/115766691-8a984400-a3a8-11eb-8007-a6ae3b3d22c7.png)

Reply:
@JoshLove-msft are you ok if I change to return list there?

Reply:
Yep, works for me - @jsquire?

Reply:
I thik it is not worth optimizing for the empty Tvalue list there or should I (it will make the code more complex and require me to have a nested class :()?

Reply:
Agreed, don't think that is worth it.

Reply:
Right now `TValue` is only ever `AmqpMap` so in theory if we remove the generic optimizing the empty case would become simpler if we want to go down that route

Reply:
Eh I don't think it is worth it - I would leave the generic.

Reply:
This blows my mind every time

![image](https://user-images.githubusercontent.com/174258/115768221-502fa680-a3aa-11eb-84f5-c42a76b3f90a.png)

the for loop squeezes out even more

Reply:
Boah with outer loop for

![image](https://user-images.githubusercontent.com/174258/115768753-03989b00-a3ab-11eb-82ef-b0736a36b6ae.png)

```
    [Config(typeof(Config))]
    public class EnumerableCount
    {
        private List<object> data;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(1600));
            }
        }
        [IterationSetup]
        public void Setup()
        {
            data = Enumerable.Repeat(new SomeClass(), NumberOfItems).Cast<object>().ToList();
        }

        [Params(0, 1, 10, 32, 50, 100, 1000)]
        public int NumberOfItems { get; set; }

        public class SomeClass {}

        [Benchmark(Baseline = true)]
        public List<SomeClass> Original()
        {
            var copyList = new List<SomeClass>();
            foreach (var value in GetValue<SomeClass>())
            {
                copyList.Add(value);
            }

            return copyList;
        }

        IEnumerable<TValue> GetValue<TValue>()
        {
            return data.Cast<TValue>();
        }

        [Benchmark]
        public List<SomeClass> List()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueList<SomeClass>();
            foreach (var value in enumerable)
            {
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        IEnumerable<TValue> GetValueList<TValue>()
        {
            List<TValue> values = null;
            foreach (var item in data)
            {
                values ??= new List<TValue>(data.Count);
                values.Add((TValue)item);
            }
            return values ?? Enumerable.Empty<TValue>();
        }

        [Benchmark]
        public List<SomeClass> ListReturnList()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueListReturnList<SomeClass>();
            foreach (var value in enumerable)
            {
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        List<TValue> GetValueListReturnList<TValue>()
        {
            var values = new List<TValue>(data.Count);
            foreach (var item in data)
            {
                values.Add((TValue)item);
            }
            return values;
        }

        [Benchmark]
        public List<SomeClass> ListReturnListFor()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueListReturnListFor<SomeClass>();
            foreach (var value in enumerable)
            {
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        [Benchmark]
        public List<SomeClass> ListForReturnListFor()
        {
            List<SomeClass> copyList = null;
            var enumerable = GetValueListReturnListFor<SomeClass>();
            for (var index = 0; index < enumerable.Count; index++)
            {
                var value = enumerable[index];
                copyList ??= enumerable is IReadOnlyCollection<SomeClass> readOnlyList
                    ? new List<SomeClass>(readOnlyList.Count)
                    : new List<SomeClass>();

                copyList.Add(value);
            }

            return copyList;
        }

        List<TValue> GetValueListReturnListFor<TValue>()
        {
            var values = new List<TValue>(data.Count);
            for (var index = 0; index < data.Count; index++)
            {
                var item = data[index];
                values.Add((TValue) item);
            }

            return values;
        }
    }
```

Reply:
I'm somewhat ambivalent on the foreach -> for change. It is nice to squeeze out more perf but I feel like we should be able to rely on language constructs and count on them to improve in perf over time. I am fine with either way here, but in general I don't think we should replace all of our foreach usages.

Reply:
For is a language construct ;) I'm aware that similar changes have been made in ASP.NET Core and the runtime as well in some places. Anyway I don't want to argue here I'm happy to go with the flow whatever you and the team prefer. I'm signing off now, so you can always edit the PR before merging

Reply:
Yeah I mean to say that it would be unfortunate if we couldn't use constructs as fundamental as foreach because of perf issues. I'm okay with the for loop here but mostly just wanted to call this out as a potential point of discussion. As always, thank you so much for the awesome contribution @danielmarbach! We greatly appreciate the time and effort you have put into improving the SDK.